xarray/core/groupby.py

```diff
@@ -294,92 +307,315 @@ def _apply_loffset(
-
-
-def _factorize_bins(
-    group, bins, cut_kwargs: Mapping | None
-) -> tuple[IndexVariable, T_GroupIndices, np.ndarray, pd.IntervalIndex, DataArray]:
-    from xarray.core.dataarray import DataArray
+@dataclass
+class ResolvedGrouper(ABC, Generic[T_Xarray]):
+    grouper: Grouper
+    group: T_Group
+    obj: T_Xarray
 
-    if cut_kwargs is None:
-        cut_kwargs = {}
-
-    if duck_array_ops.isnull(bins).all():
-        raise ValueError("All bin edges are NaN.")
-    binned, bins = pd.cut(group.values, bins, **cut_kwargs, retbins=True)
-    codes = binned.codes
-    if (codes == -1).all():
-        raise ValueError(f"None of the data falls within bins with edges {bins!r}")
-    full_index = binned.categories
-    uniques = np.sort(pd.unique(codes))
-    unique_values = full_index[uniques[uniques != -1]]
-    group_indices = [g for g in _codes_to_groups(codes, len(full_index)) if g]
-
-    if len(group_indices) == 0:
-        raise ValueError(f"None of the data falls within bins with edges {bins!r}")
-
-    new_dim_name = str(group.name) + "_bins"
-    group_ = DataArray(binned, getattr(group, "coords", None), name=new_dim_name)
-    unique_coord = IndexVariable(new_dim_name, unique_values)
-    return unique_coord, group_indices, codes, full_index, group_
-
-
-def _factorize_rest(
-    group,
-) -> tuple[IndexVariable, T_GroupIndices, np.ndarray]:
-    # look through group to find the unique values
-    group_as_index = safe_cast_to_index(group)
-    sort = not isinstance(group_as_index, pd.MultiIndex)
-    unique_values, group_indices, codes = unique_value_groups(group_as_index, sort=sort)
-    if len(group_indices) == 0:
-        raise ValueError(
-            "Failed to group data. Are you grouping by a variable that is all NaN?"
+    _group_as_index: pd.Index | None = field(default=None, init=False)
+
+    # Defined by factorize:
+    labels: Any | None = field(default=None, init=False)  # TODO: Typing?
+    codes: DataArray = field(init=False)
+    group_indices: T_GroupIndices = field(init=False)
+    unique_coord: IndexVariable | _DummyGroup = field(init=False)
+    full_index: pd.Index = field(init=False)
+
+    # _ensure_1d:
+    group1d: T_Group = field(init=False)
+    stacked_obj: T_Xarray = field(init=False)
+    stacked_dim: Hashable | None = field(init=False)
+    inserted_dims: list[Hashable] = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.group: T_Group = _resolve_group(self.obj, self.group)
+
+        (
+            self.group1d,
+            self.stacked_obj,
+            self.stacked_dim,
+            self.inserted_dims,
+        ) = _ensure_1d(group=self.group, obj=self.obj)
+
+    @property
+    def name(self) -> Hashable:
+        return self.group1d.name
+
+    @property
+    def size(self) -> int:
+        return len(self)
+
+    def __len__(self) -> int:
+        return len(self.full_index)  # TODO: full_index not def, abstractmethod?
```

Comment:
This will crash if `.factorize` hasn't been triggered before.

Reply:
Yes but it wouldn't make sense without it, and this class is internal-only